### PR TITLE
[inductor] Add torch.while_loop support to JIT Inductor

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122069

Summary: `torch.while_loop` HOP support is added to JIT Inductor. The test coverage is limited due to the functionality constraints of the upstream `torch.while_loop` op in Dynamo / Export. When those are lifted, we'll add more tests (see TODO-s in the test file).

AOT Inductor support will be added in a follow-up PR.

Test Plan:

```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 38 tests in 159.387s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -7138,6 +7139,111 @@ def codegen(self, wrapper):
...
+        assert (
+            len(operands) > 0
+        ), "torch.while_loop is assumed to have at least one operand."
+
+        while_loop = WhileLoop(
+            operands=operands,
+            cond_subgraph=cond_fn,
+            body_subgraph=body_fn,
+            # asserted above that there is at least one operand
+            layout=MultiOutputLayout(operands[0].get_device()),
```
Comment (eellison) [inline torch/_inductor/ir.py:7222]: What happens if there are multiple devices ?

**Review by eellison — APPROVED** [review]
looks good!

File: torch/_inductor/ir.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/ir.py:7222]: Good question. I guess, for now we can restrict all the `operands` to be on the same device, as `MultiOutputLayout` assumes a single device.

**Pull request closed without merging**